pkg/virt-handler/vm.go

```diff
@@ -291,16 +291,32 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 				domainInterfaceStatusByMac[domainInterfaceStatus.Mac] = domainInterfaceStatus
 			}
 
+			existingInterfacesSpecByName := map[string]v1.Interface{}
+			for _, existingInterfaceSpec := range vmi.Spec.Domain.Devices.Interfaces {
+				if existingInterfaceSpec.Name != "" {
+					existingInterfacesSpecByName[existingInterfaceSpec.Name] = existingInterfaceSpec
+				}
+			}
+
 			// Iterate through all domain.Spec interfaces
 			for _, domainInterface := range domain.Spec.Devices.Interfaces {
 				interfaceMAC := domainInterface.MAC.MAC
 				var newInterface v1.VirtualMachineInstanceNetworkInterface
+				var isMasqueredeInterface bool = false
+
+				if existingInterfacesSpecByName[domainInterface.Alias.Name].Masquerade != nil {
+					isMasqueredeInterface = true
+				}
 
 				if existingInterface, exists := existingInterfaceStatusByName[domainInterface.Alias.Name]; exists {
 					// Reuse previously calculated interface from vmi.Status.Interfaces, updating the MAC from domain.Spec
 					// Only interfaces defined in domain.Spec are handled here
 					newInterface = existingInterface
-					newInterface.MAC = interfaceMAC
+					// Do not update if interface has Masquerade binding
+					// virt-controller should update VMI status interface instead
+					if !isMasqueredeInterface {
+						newInterface.MAC = interfaceMAC
```

Comment:
It is clear that for IP we want to expose the one externally accessible. I'm not so sure what to do with MAC though. Should we report the MAC address that is used inside the guest or the one that is on the virt-launcher pod?

When a user explicitly specifies a MAC address for their VM, I suspect they would expect that MAC to appear on the VMI report.

Does that sound right @rmohr ?

Reply:
@rmohr @phoracek @vatsalparekh If a custom MAC was configured on the VM, should it appear in the VMI spec? if so, we can try to locate it and if it exist we won't overwrite it but just compare it.

Reply:
I believe so. As a user, I'd expect to see the internal MAC used in the VM.

I don't understand what you mean by "if so, we can try to locate it and if it exist we won't overwrite it but just compare it.".

Reply:
I see a few options:
1. Expose external facing details (public IP and public MAC)
2. Expose internal details (vNIC IP and vNIC MAC)
3. Expose external IP and vNIC MAC
4. Expose vNIC IP and external MAC
5. Expose external and internal (vNIC) details on the status

TO me 4 does not make much sense, but we shoul dunderstand which use-cases are supported by options 1-3. We should ask ourselves: When is the MAC needed and when i the IP needed.
Why do we expose these informations on the API, and thus which of the options above does answer those user questions.

/cc @vladikr 
xref #3090

Reply:
@fabiand 
In my view, the IPs in the status are there to reflect what IPs does the guest sees. This also serves as an indication to the user that the guest managed to get an IP(s).
For me, `status.IPs` is not about how the API user can connect to the VMI.
I don't think that all networks that the guest will be able to connect to will also be reachable by the API user.

Therefore, I'm not in favor of this PR. :/

Wrt #3090 , this is a completely different use case, from my POV.  This is about the guest OS user being about to identify which vNIC is connected to which network. In some cases, it is needed to configure the interface properly or to configure a vlan or anything else, but the guest user needs to see the vNIC information to be able to identify it.

Reply:
@vladikr I, as a user, would expect to publicly accessible IP on the VMI. Seeing guest agent report about the inside configuration may be interesting, but I find it way more useful to be able to get the public IP through the VMI. Then I can use it to create my own Endpoints/Services, connect to the running VM directly from another one etc.

Also, the default now is to report the Pod IP. I consider the report done by guest-agent something extra. If we want to make the report stable with or without guest-agent, using the default (pod IP) would be less destructive to the API.

Reply:
@vladikr 

We have two bindings, `bridge` and `masquerade`. We also have two modes of reporting IP, with and without guest-agent.

With `bridge`, it is simple. Since we move the Pod IP inside the VM and extend the Pod's L2 network to the VM, what guest-agent sees in the VM matches the Pod IP. No problem here.

Now with `masquerade`, we keep the original IP in the Pod. The VM is connected through NAT and keeps a link-local IP which gets traffic forwarded from the Pod. If we don't gave guest-agent running, we report the Pod IP. With that, other Pods/VMs can use this IP to access the VM. However, when guest-agent is present, we report the internal link-local IP. That has two issues: 1) Reported IP is not accessible from the outside and 2) The reported IP changes during the booting, first we report Pod IP, once guest-agent kicks in, we start reporting the link-local.

Mission of this PR is to adjust reporting of IP when using `masquerade` so we always report the Pod IP, even once guest-agent starts. That way we always report the publicly accessible IP and we keep the reporting consistent with or without guest-agent.

Please note that reporting of the link-local would have little to no use since users cannot change it after the VM starts (that would break forwarding).